In [1]:
from sklearn import datasets, svm, metrics
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt

digits = datasets.load_digits()

In [2]:
#b
n_samples = len(digits.images)
data = digits.images.reshape((n_samples, -1))
clf =svm.SVC(gamma=0.001)
X_train, X_test, y_train, y_test = train_test_split(data, digits.target, test_size=0.2)

In [3]:
#c
from sklearn.model_selection import KFold, cross_val_score

X, y = datasets.load_digits(return_X_y=True)
k_folds = KFold(n_splits = 5)
scores = cross_val_score(clf, X, y, cv = k_folds)

print("Cross Validation:", scores)

Cross Validation: [0.975      0.95277778 0.98328691 0.99164345 0.95821727]


In [4]:
#d
param_grid = { 
  'C': [0.1, 1, 10, 100, 1000], 
  'gamma': [1, 0.1, 0.01, 0.001, 0.0001] 
}

In [5]:
from sklearn.model_selection import GridSearchCV 
from sklearn.svm import SVC

grid=GridSearchCV(SVC(), param_grid, refit=True, verbose=3)
grid.fit(X_train,y_train)

Fitting 5 folds for each of 25 candidates, totalling 125 fits
[CV 1/5] END ....................C=0.1, gamma=1;, score=0.108 total time=   1.1s
[CV 2/5] END ....................C=0.1, gamma=1;, score=0.104 total time=   1.0s
[CV 3/5] END ....................C=0.1, gamma=1;, score=0.105 total time=   1.0s
[CV 4/5] END ....................C=0.1, gamma=1;, score=0.105 total time=   1.1s
[CV 5/5] END ....................C=0.1, gamma=1;, score=0.105 total time=   1.0s
[CV 1/5] END ..................C=0.1, gamma=0.1;, score=0.108 total time=   1.1s
[CV 2/5] END ..................C=0.1, gamma=0.1;, score=0.104 total time=   0.9s
[CV 3/5] END ..................C=0.1, gamma=0.1;, score=0.105 total time=   0.9s
[CV 4/5] END ..................C=0.1, gamma=0.1;, score=0.105 total time=   0.9s
[CV 5/5] END ..................C=0.1, gamma=0.1;, score=0.105 total time=   0.8s
[CV 1/5] END .................C=0.1, gamma=0.01;, score=0.205 total time=   0.6s
[CV 2/5] END .................C=0.1, gamma=0.01

[CV 2/5] END ...................C=1000, gamma=1;, score=0.104 total time=   1.0s
[CV 3/5] END ...................C=1000, gamma=1;, score=0.105 total time=   1.1s
[CV 4/5] END ...................C=1000, gamma=1;, score=0.105 total time=   1.0s
[CV 5/5] END ...................C=1000, gamma=1;, score=0.105 total time=   1.0s
[CV 1/5] END .................C=1000, gamma=0.1;, score=0.115 total time=   0.9s
[CV 2/5] END .................C=1000, gamma=0.1;, score=0.108 total time=   0.9s
[CV 3/5] END .................C=1000, gamma=0.1;, score=0.105 total time=   0.9s
[CV 4/5] END .................C=1000, gamma=0.1;, score=0.105 total time=   0.9s
[CV 5/5] END .................C=1000, gamma=0.1;, score=0.108 total time=   0.9s
[CV 1/5] END ................C=1000, gamma=0.01;, score=0.840 total time=   0.6s
[CV 2/5] END ................C=1000, gamma=0.01;, score=0.795 total time=   0.6s
[CV 3/5] END ................C=1000, gamma=0.01;, score=0.770 total time=   0.6s
[CV 4/5] END ...............

GridSearchCV(estimator=SVC(),
             param_grid={'C': [0.1, 1, 10, 100, 1000],
                         'gamma': [1, 0.1, 0.01, 0.001, 0.0001]},
             verbose=3)

In [6]:
grid.best_params_

{'C': 10, 'gamma': 0.001}

In [7]:
grid.best_estimator_

SVC(C=10, gamma=0.001)

In [8]:
grid_predictions=grid.predict(X_test)

In [9]:
from sklearn.metrics import classification_report,confusion_matrix

print(confusion_matrix(y_test,grid_predictions))

[[41  0  0  0  0  0  0  0  0  0]
 [ 0 33  0  0  0  0  0  0  0  0]
 [ 0  0 31  0  0  0  0  0  0  0]
 [ 0  0  0 35  0  0  0  0  0  1]
 [ 0  0  0  0 47  0  0  0  0  0]
 [ 0  0  0  0  0 33  0  0  0  1]
 [ 0  0  0  0  0  0 30  0  0  0]
 [ 0  0  0  0  0  0  0 31  0  0]
 [ 0  1  0  0  0  0  0  0 34  2]
 [ 0  0  0  0  0  1  0  0  0 39]]


In [10]:
print(classification_report(y_test,grid_predictions))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        41
           1       0.97      1.00      0.99        33
           2       1.00      1.00      1.00        31
           3       1.00      0.97      0.99        36
           4       1.00      1.00      1.00        47
           5       0.97      0.97      0.97        34
           6       1.00      1.00      1.00        30
           7       1.00      1.00      1.00        31
           8       1.00      0.92      0.96        37
           9       0.91      0.97      0.94        40

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



2.
a.A acurácia é de 98%, é dada pela divisão do número de acertos e o total de entradas.

b.Sim, pois a porcentagem de precisão, recall e f1 ficaram próximas de 1.

c.{'C': 10, 'gamma': 0.001}

d.Banco de dados reais são suscetíveis a ruídos, ausência e inconsistência de dados devido ao seu tamanho, geralmente muito grande, e provavelmente, origem de múltiplas fontes heterogêneas.
Deve-se sempre alinhar os times de ciência de dados e negócios para ter o entendimento de qual modelo é o mais adequado para o negócio em questão, qual das métricas será mais adequada para cada caso.

3.
a.Um parâmetro de modelo é uma variável do modelo selecionado que pode ser estimada ajustando os dados fornecidos ao modelo. Um hiperparâmetro de modelo é o parâmetro cujo valor é definido antes de o modelo iniciar o treinamento. Eles não podem ser aprendidos ajustando o modelo aos dados.

b. 1-Entender Quais Processos Precisam de Automação. Os processos mais fáceis de automatizar são aqueles que são realizados manualmente a cada dia sem resultado variável. Processos complicados requerem mais introspecção antes da automação. Enquanto o aprendizado de máquina pode definitivamente ajudar a automatizar alguns processos, nem todos os problemas de automação precisam de aprendizado de máquina. 
2-Começando Sem Bons Dados. O problema número 1 enfrentado durante a aprendizagem de máquina começa com a falta de bons dados. Embora o aprimoramento de algoritmos muitas vezes consome a maior parte do tempo em Inteligencia Artificial (AI), a qualidade dos dados é essencial para que os algoritmos funcionem conforme pretendido. Dados danificados, dados corrompidos e dados incompletos são os inimigos essenciais do aprendizado de máquina ideal.
3-Infraestrutura Inadequada. O aprendizado de máquina requer grandes quantidades de recursos de manipulacao de dados. Sistemas herdados muitas vezes não podem lidar com a carga de trabalho e se dobram sob pressão. Você deve verificar se sua infraestrutura pode lidar com aprendizado de máquina. Se não puder, você deve procurar atualizar, concluir com aceleração de hardware e armazenamento flexível. 
4-Implementação. As organizações geralmente têm mecanismos de análise trabalhando com eles no momento em que optam por atualizar para o aprendizado de máquina. Integrar metodologias de aprendizado de máquina mais recentes em metodologias existentes é uma tarefa complicada.

c.É provável que o modelo esteja sobrescrevendo os dados de treinamento (ou tivemos muita sorte com os dados de treinamento). As possíveis soluções para o overfitting é basicamente obter mais dados e simplificando o modelo. (selecionando um algoritmo mais simples, reduzindo o número de parâmetros ou recursos utilizados ou regularizando o modelo) ou reduzindo o ruído nos dados de treinamento.
Soluções: 
treinar com mais dados-Se a máquina de aprendizagem usada é complexa, em termos da quantidade de parâmetros a ajustar, uma alternativa é adquirir mais dados com o intuito de equilibrar a quantidade de parâmetros versus a quantidade de instâncias de treinamento.
validação cruzada- Uma das formas de realizar validação cruzada é usar o procedimento k-fold cross-validation. Nesse procedimento, o conjunto de dados é dividido em k partes, aproximadamente do mesmo tamanho, das quais, k-1 partes são usadas para treinar o modelo e a parte restante é usada para avaliar o modelo. Esse processo é repetido k vezes, de forma que cada parte será usada tanto para treinar como para avaliar o modelo. De maneira geral, a validação cruzada, por si só, não evita overfitting, mas segue uma boa prática ao separar o conjunto de teste e ao realizar um revezamento dos dados para uma melhor avaliação, no que tange a generalização do modelo em instâncias não vistas. Uma observação: o k-fold cross-validation não é uma boa opção quando o conjunto possui poucos dados.
parar o treinamento mais cedo (early stopping)- Máquinas de aprendizagem, tais como redes neurais artificiais, árvores de decisão, deep learning, entre outras, aprendem iterativamente. A cada passo, a máquina ajusta seus parâmetros aos dados e isso pode ser monitorado. Pode-se usar esse monitoramento para decidir qual é o melhor momento de interromper o treinamento da máquina. Espera-se que a precisão no conjunto de treinamento aumente com o tempo, mas, em relação ao conjunto de validação, a acurácia deve atingir um pico e depois cair. Esse pode ser um bom momento para frear o treinamento, antes que a máquina se sobreajuste aos dados.

d.Um conjunto de teste é usado para estimar o erro de generalização que um modelo fará em novas instâncias, antes que o modelo seja lançado na produção.

e.Um conjunto de validação é usado para comparar modelos. Permite selecionar o melhor modelo e ajustar os hiperparâmetros.

f.Se você ajustar os hiperparâmetros usando o conjunto de testes, corre o risco de superajustá-lo, e o erro de generalização que você medir será otimista(você pode iniciar um modelo com desempenho pior do que o esperado)